File: packages/common/utils/math/Coordinates/PolarOffset.ts
import { Generics } from '../Generics';
import { IOffset } from '../IOffset';
import { Numbers } from '../Numbers';
import { Angle } from './Angle';
import { CartesianOffset } from './CartesianOffset';
import { PolarCoordinate } from './PolarCoordinate';

/**
 * Represents the difference between polar coordinates I (first) and J (second) in two-dimensional space.
 * @date 2/3/2024 - 9:04:58 AM
 *
 * @see {@link https://en.wikipedia.org/wiki/Polar_coordinate_system}
 * @export
 * @class PolarOffset
 * @typedef {PolarOffset}
 * @implements {IOffset<PolarCoordinate>}
 */
export class PolarOffset implements IOffset<PolarCoordinate> {

  readonly Tolerance: number;
  readonly I: PolarCoordinate;
  readonly J: PolarCoordinate;

  /**
   * 𝝙r = rj - ri.
   *
   * @readonly
   * @type {number}
   */
  get radius(): number {
    return (this.J.radius - this.I.radius);
  }

  /**
   * 𝝙φ = φj - φi.
   *
   * @readonly
   * @type {Angle}
   */
  get azimuth(): Angle {
    return this.J.azimuth.subtractBy(this.I.azimuth);
  }

  // ==== Initialization
  /**
   * Creates an instance of PolarOffset.
   * @date 2/3/2024 - 9:04:58 AM
   *
   * @constructor
   * @param {PolarCoordinate} i The first coordinate.
   * @param {PolarCoordinate} j The second coordinate.
   * @param {number} [tolerance=Numbers.ZeroTolerance]
   */
  protected constructor(
    i: PolarCoordinate,
    j: PolarCoordinate,
    tolerance: number = Numbers.ZeroTolerance) {
    this.I = i;
    this.J = j;
    this.Tolerance = tolerance;
  }

  /**
   * Creates an instance of CartesianOffset from the provided coordinates.
   * @date 2/3/2024 - 9:04:58 AM
   *
   * @static
   * @param {PolarCoordinate} i
   * @param {PolarCoordinate} j
   * @param {number} [tolerance=Numbers.ZeroTolerance]
   * @returns {PolarOffset}
   */
  static fromCoordinates(
    i: PolarCoordinate,
    j: PolarCoordinate,
    tolerance: number = Numbers.ZeroTolerance
  ): PolarOffset {
    return new PolarOffset(i, j, tolerance)
  }

  /**
   * Creates an instance of PolarOffset from the specified offsets from the global origin.
   *
   * @static
   * @param {number} deltaRadius The radius offset from the origin.
   * @param {number} deltaAzimuth The rotation offset from the origin.
   * @param {number} [tolerance=Numbers.ZeroTolerance]
   * @return {*}  {CartesianOffset}
   * @memberof CartesianOffset
   */
  static fromOffsets(
    deltaRadius: number,
    deltaAzimuth: Angle,
    tolerance: number = Numbers.ZeroTolerance
  ): PolarOffset {
    const i = PolarCoordinate.atOrigin();
    const j = PolarCoordinate.fromAngle(deltaRadius, deltaAzimuth, tolerance);

    return new PolarOffset(i, j, tolerance);
  }

  // ==== Conversions
  toString(): string {
    return `${PolarOffset.name} - `
      + `I: (r:${this.I.radius.toFixed(6)}, a:${this.I.azimuth.Radians.toFixed(6)}), `
      + `J: (r:${this.J.radius.toFixed(6)}, a:${this.J.azimuth.Radians.toFixed(6)})`;
  }

  /**
    * Converts to a single coordinate with the global origin aligned with the first point of the offset.
    *
    * @return {*}  {CartesianCoordinate}
    * @memberof CartesianOffset
    */
  toPolarCoordinate(): PolarCoordinate {
    return PolarCoordinate.fromAngle(
      this.radius,
      this.azimuth,
      this.Tolerance);
  }

  /**
   * Converts to Cartesian offset.
   * @returns {CartesianOffset} CartesianOffset.
   */
  toCartesianOffset(): CartesianOffset {
    return CartesianOffset.fromCoordinates(this.I.toCartesian(), this.J.toCartesian(), this.Tolerance);
  }

  // ==== Public Methods
  /**
   * Calculates the total straight length between the offset points.
   * @returns {number} The total straight length.
   * @see {@link https://www.ck12.org/book/ck-12-trigonometry-concepts/section/6.2/}
   */
  arcLength(): number {
    return Math.sqrt(
      this.I.radius ** 2
      + this.J.radius ** 2
      - 2 * this.I.radius * this.J.radius * (Math.cos(this.azimuth.Radians))
    );
  }

  // TODO: Finish chordSeparation
  // chordSeparation(): number {

  // }

  // ==== Comparisons
  /**
   * Indicates whether the current object is equal to another object of the same type.
   * @param {PolarOffset} other An object to compare with this object.
   * @returns {boolean} true if the current object is equal to the other parameter; otherwise, false.
   */
  equals(other: PolarOffset): boolean {
    const tolerance: number = Math.min(this.Tolerance, other.Tolerance);
    return (
      Numbers.AreEqual(this.radius, other.radius, tolerance) &&
      Numbers.AreEqual(this.azimuth.Radians, other.azimuth.Radians, tolerance)
    );
  }

  // ==== Combinations
  addToCoordinate(coord: PolarCoordinate): PolarCoordinate {
    return PolarCoordinate.fromAngle(
      this.radius + coord.radius,
      coord.azimuth.addTo(this.azimuth),
      Math.max(coord.Tolerance, this.Tolerance)
    );
  }

  subtractFromCoordinate(coord: PolarCoordinate): PolarCoordinate {
    return PolarCoordinate.fromAngle(
      coord.radius - this.radius,
      coord.azimuth.subtractBy(this.azimuth),
      Math.max(coord.Tolerance, this.Tolerance)
    );
  }

  subtractByCoordinate(coord: PolarCoordinate): PolarCoordinate {
    return PolarCoordinate.fromAngle(
      this.radius - coord.radius,
      this.azimuth.subtractBy(coord.azimuth),
      Math.max(coord.Tolerance, this.Tolerance)
    );
  }

  addTo(offset: PolarOffset): PolarOffset {
    return new PolarOffset(
      PolarCoordinate.atOrigin(),
      PolarCoordinate.fromAngle(
        this.radius + offset.radius,
        this.azimuth.addTo(offset.azimuth)
      ),
      Generics.getToleranceBetween(this, offset)
    );
  }

  subtractBy(offset: PolarOffset): PolarOffset {
    return new PolarOffset(
      PolarCoordinate.atOrigin(),
      PolarCoordinate.fromAngle(
        this.radius + offset.radius,
        this.azimuth.subtractBy(offset.azimuth)
      ),
      Generics.getToleranceBetween(this, offset)
    );
  }

  multiplyBy(multiplier: number): PolarOffset {
    return new PolarOffset(
      this.I.multiplyBy(multiplier),
      this.J.multiplyBy(multiplier),
      this.Tolerance
    );
  }

  divideBy(denominator: number): PolarOffset {
    return new PolarOffset(
      this.I.divideBy(denominator),
      this.J.divideBy(denominator),
      this.Tolerance
    );
  }
}

File: packages/libraries/math/src/Coordinates/PolarOffset.cs
using MPT.Math.NumberTypeExtensions;
using System;
using NMath = System.Math;

namespace MPT.Math.Coordinates
{
    /// <summary>
    /// Represents the difference between polar coordinates I (first) and J (second) in two-dimensional space.
    /// </summary>
    /// <seealso cref="System.IEquatable{PolarOffset}" />
    public struct PolarOffset : IEquatable<PolarOffset>, ITolerance
    {
        #region Properties
        /// <summary>
        /// Tolerance to use in all calculations with double types.
        /// </summary>
        /// <value>The tolerance.</value>
        public double Tolerance { get; set; }

        /// <summary>
        /// Gets or sets the first coordinate value.
        /// </summary>
        /// <value>The first coordinate.</value>
        public PolarCoordinate I { get; private set; }

        /// <summary>
        /// Gets or sets the second coordinate value.
        /// </summary>
        /// <value>The second coordinate.</value>
        public PolarCoordinate J { get; private set; }
        #endregion

        #region Initialization
        /// <summary>
        /// Initializes a new instance of the <see cref="CartesianOffset"/> struct.
        /// </summary>
        /// <param name="i">The first coordinate.</param>
        /// <param name="j">The second coordinate.</param>
        /// <param name="tolerance">The tolerance.</param>
        public PolarOffset(
            PolarCoordinate i, PolarCoordinate j,
            double tolerance = Numbers.ZeroTolerance)
        {
            I = i;
            J = j;
            Tolerance = tolerance;
        }
        #endregion

        #region Methods: Public
        /// <summary>
        /// Returns a <see cref="System.String" /> that represents this instance.
        /// </summary>
        /// <returns>A <see cref="System.String" /> that represents this instance.</returns>
        public override string ToString()
        {
            return base.ToString() +
                " - I: (r:" + NMath.Round(I.Radius, 6) + ", a:" + NMath.Round(I.Azimuth.Radians, 6)
                + "), J: (r:" + NMath.Round(J.Radius, 6) + ", a:" + NMath.Round(J.Azimuth.Radians, 6) + ")";
        }

        /// <summary>
        /// Converts to cartesian offset.
        /// </summary>
        /// <returns>CartesianOffset.</returns>
        public CartesianOffset ToCartesian()
        {
            return new CartesianOffset(I, J, Tolerance);
        }

        /// <summary>
        /// Radius_j - Radius_i.
        /// </summary>
        /// <returns>System.Double.</returns>
        public double Radius()
        {
            return (J.Radius - I.Radius);
        }

        /// <summary>
        /// Azimuth_j - Azimuth_i.
        /// </summary>
        /// <returns>System.Double.</returns>
        public Angle Azimuth()
        {
            return (J.Azimuth - I.Azimuth);
        }

        /// <summary>
        /// The total straight length between the offset points.
        /// <seealso ref="https://www.ck12.org/book/ck-12-trigonometry-concepts/section/6.2/"/>
        /// </summary>
        /// <returns>System.Double.</returns>
        public double Length()
        {
            return NMath.Sqrt(I.Radius.Squared() + J.Radius.Squared() - 2 * I.Radius * J.Radius * (NMath.Cos(J.Azimuth.Radians - I.Azimuth.Radians)));
        }
        #endregion

        #region Operators & Equals
        /// <summary>
        /// Indicates whether the current object is equal to another object of the same type.
        /// </summary>
        /// <param name="other">An object to compare with this object.</param>
        /// <returns>true if the current object is equal to the <paramref name="other" /> parameter; otherwise, false.</returns>
        public bool Equals(PolarOffset other)
        {
            double tolerance = NMath.Min(Tolerance, other.Tolerance);
            return Radius().IsEqualTo(other.Radius(), tolerance) &&
                   Azimuth().Radians.IsEqualTo(other.Azimuth().Radians, tolerance);
        }

        /// <summary>
        /// Determines whether the specified <see cref="System.Object" /> is equal to this instance.
        /// </summary>
        /// <param name="obj">The object to compare with the current instance.</param>
        /// <returns><c>true</c> if the specified <see cref="System.Object" /> is equal to this instance; otherwise, <c>false</c>.</returns>
        public override bool Equals(object obj)
        {
            if (obj is PolarOffset) { return Equals((PolarOffset)obj); }
            return base.Equals(obj);
        }

        /// <summary>
        /// Returns a hash code for this instance.
        /// </summary>
        /// <returns>A hash code for this instance, suitable for use in hashing algorithms and data structures like a hash table.</returns>
        public override int GetHashCode()
        {
            return I.GetHashCode() ^ J.GetHashCode();
        }


        /// <summary>
        /// Implements the == operator.
        /// </summary>
        /// <param name="a">Offset a.</param>
        /// <param name="b">Offset b.</param>
        /// <returns>The result of the operator.</returns>
        public static bool operator ==(PolarOffset a, PolarOffset b)
        {
            return a.Equals(b);
        }
        /// <summary>
        /// Implements the != operator.
        /// </summary>
        /// <param name="a">Offset a.</param>
        /// <param name="b">Offset b.</param>
        /// <returns>The result of the operator.</returns>
        public static bool operator !=(PolarOffset a, PolarOffset b)
        {
            return !a.Equals(b);
        }

        /// <summary>
        /// Implements the - operator.
        /// </summary>
        /// <param name="point1">The point1.</param>
        /// <param name="offset2">The offset2.</param>
        /// <returns>The result of the operator.</returns>
        public static PolarCoordinate operator -(PolarCoordinate point1, PolarOffset offset2)
        {
            return new PolarCoordinate(
                point1.Radius - offset2.Radius(),
                point1.Azimuth - offset2.Azimuth(),
                NMath.Max(point1.Tolerance, offset2.Tolerance));
        }
        /// <summary>
        /// Implements the - operator.
        /// </summary>
        /// <param name="offset1">The offset1.</param>
        /// <param name="point2">The point2.</param>
        /// <returns>The result of the operator.</returns>
        public static PolarCoordinate operator -(PolarOffset offset1, PolarCoordinate point2)
        {
            return offset1 + (-1 * point2);
        }

        /// <summary>
        /// Implements the + operator.
        /// </summary>
        /// <param name="point1">The point1.</param>
        /// <param name="offset2">The offset2.</param>
        /// <returns>The result of the operator.</returns>
        public static PolarCoordinate operator +(PolarCoordinate point1, PolarOffset offset2)
        {
            return offset2 + point1;
        }
        /// <summary>
        /// Implements the + operator.
        /// </summary>
        /// <param name="offset1">The offset1.</param>
        /// <param name="point2">The point2.</param>
        /// <returns>The result of the operator.</returns>
        public static PolarCoordinate operator +(PolarOffset offset1, PolarCoordinate point2)
        {
            return new PolarCoordinate(
                offset1.Radius() + point2.Radius,
                offset1.Azimuth() + point2.Azimuth,
                NMath.Max(offset1.Tolerance, point2.Tolerance));
        }


        /// <summary>
        /// Implements the * operator.
        /// </summary>
        /// <param name="offset">The offset.</param>
        /// <param name="multiplier">The multiplier.</param>
        /// <returns>The result of the operator.</returns>
        public static PolarOffset operator *(PolarOffset offset, double multiplier)
        {
            return multiplier * offset;
        }

        /// <summary>
        /// Implements the * operator.
        /// </summary>
        /// <param name="multiplier">The multiplier.</param>
        /// <param name="offset">The offset.</param>
        /// <returns>The result of the operator.</returns>
        public static PolarOffset operator *(double multiplier, PolarOffset offset)
        {
            return new PolarOffset(
                offset.I * multiplier,
                offset.J * multiplier,
                offset.Tolerance);
        }


        /// <summary>
        /// Implements the / operator.
        /// </summary>
        /// <param name="offset">The offset.</param>
        /// <param name="denominator">The denominator.</param>
        /// <returns>The result of the operator.</returns>
        public static PolarOffset operator /(PolarOffset offset, double denominator)
        {
            return new PolarOffset(
                offset.I / denominator,
                offset.J / denominator,
                offset.Tolerance);
        }
        #endregion

        #region Operators: Conversion
        /// <summary>
        /// Performs an implicit conversion from <see cref="PolarOffset"/> to <see cref="CartesianOffset"/>.
        /// </summary>
        /// <param name="polar">The polar.</param>
        /// <returns>The result of the conversion.</returns>
        public static implicit operator CartesianOffset(PolarOffset polar) => polar.ToCartesian();
        #endregion
    }
}